libsolidity/codegen/ExpressionCompiler.cpp

```diff
@@ -1249,21 +1251,56 @@ bool ExpressionCompiler::visit(FunctionCall const& _functionCall)
 		case FunctionType::Kind::Require:
 		{
 			acceptAndConvert(*arguments.front(), *function.parameterTypes().front(), false);
-
+			// stack: <condition>
 			bool haveReasonString = arguments.size() > 1 && m_context.revertStrings() != RevertStrings::Strip;
-
 			if (arguments.size() > 1)
 			{
 				// Users probably expect the second argument to be evaluated
 				// even if the condition is false, as would be the case for an actual
 				// function call.
 				solAssert(arguments.size() == 2, "");
 				solAssert(function.kind() == FunctionType::Kind::Require, "");
-				// Check if the function call matches ``require(bool, error)``, and throw an unimplemented error,
-				// as require with custom errors is not supported in legacy codegen.
 				auto const* magicType = dynamic_cast<MagicType const*>(arguments[1]->annotation().type);
 				if (magicType && magicType->kind() == MagicType::Kind::Error)
-					solUnimplemented("Require with a custom error is only available using the via-ir pipeline.");
+				{
+					// Make sure that error constructor arguments are evaluated regardless of the require condition
+					auto const& errorConstructorCall = dynamic_cast<FunctionCall const&>(*arguments[1]);
+					errorConstructorCall.expression().accept(*this);
+					std::vector<Type const*> errorConstructorArgumentTypes{};
+					for (ASTPointer<Expression const> const& errorConstructorArgument: errorConstructorCall.sortedArguments())
+					{
+						errorConstructorArgument->accept(*this);
+						errorConstructorArgumentTypes.push_back(errorConstructorArgument->annotation().type);
+					}
+					unsigned const sizeOfConditionArgument = arguments.at(0)->annotation().type->sizeOnStack();
+					unsigned const sizeOfErrorArguments = CompilerUtils::sizeOnStack(errorConstructorArgumentTypes);
+					if (sizeOfErrorArguments + sizeOfConditionArgument > 16)
+						BOOST_THROW_EXCEPTION(
+							StackTooDeepError() <<
+							errinfo_sourceLocation(errorConstructorCall.location()) <<
+							util::errinfo_comment(util::stackTooDeepString)
+						);
```

Comment:
Ah, even better if that's the case, since for most other occurrences where we throw a stack too deep, I saw that we were manually checking the offset.

Reply:
Wait, for the latter case (`InvalidOpcode`) - I don't think you saw correctly? If the stack offset goes beyond what is supported, the assertion that throws `StackToDeep` will throw before we actually even try to rotate the stack via swaps; I assume `InvalidOpcode` would be thrown if we tried to call `SWAP17`?

Reply:
I'll use the first suggestion (and rethrow).

Reply:
> Wait, for the latter case (`InvalidOpcode`) - I don't think you saw correctly?

You may be right. I meant whichever exception is thrown as a result of the test case from https://github.com/ethereum/solidity/pull/15174#discussion_r1640116564.

I remembered it to be `StackTooDeep`, but when I was making this comment I quickly looked into `moveToStackTop()` and didn't see anything around it throwing that so I assumed it must have actually been `InvalidOpcode` thrown when we try to swap too far. But I guess you're saying that there's some assert that throws `StackTooDeep` before that happens?

In any case, if you checked with that test case and the exception is now caught and properly rethrown, it's fine.